scipy/interpolate/_bspl.pyx

```diff
@@ -412,3 +413,52 @@ def _norm_eq_lsq(const double[::1] x,
                 # ... and A.T @ y
                 for ci in range(rhs.shape[1]):
                     rhs[row, ci] = rhs[row, ci] + wrk[r] * y[j, ci] * wval
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def _make_design_matrix(const double[::1] x,
+                const double[::1] t,
+                int k,
+                kind):
+    '''
+    Returns a design matrix for BSpline object in "dense" or "CSR" formats.
+    
+    Parameters
+    ----------
+    x : 1-D array, shape(n,)
+        Values of x - coordinate of a given set of points (n >= 1).   
+    t : 1-D array, shape(nt,)
+        Vector of knots.
+    k : int
+        The maximum degree of spline.
+    kind : "CSR", optional
+        Default is ``"CSR"`` .
+        * ``"CSR"`` : Returns sparse matrix in CSR.
+
+    Returns
+    -------
+    Sparse matrix in CSR.
+    '''
+    n = len(x)
+    nt = len(t)
```

Comment:
```suggestion
    cdef:
        int i, ind
        int n = len(x)
        int nt = len(t)
        double[::1] data = np.zeros(n * (k + 1), dtype=np.float64)
        int[::1] row_ind = np.zeros(n * (k + 1), dtype=np.int64)
        int[::1] col_ind = np.zeros(n * (k + 1), dtype=np.int64)
```
I'm not entirely sure about the dtypes.
The reasoning behind setting memoryviews like `double[::1] data` is that this does not need Python interaction, but Cython can work in C alone, which can potentially speed up the runtime.

Reply:
Integer dtypes should probably be np.int64_t, for portability.